packages/python/readme_metrics/PayloadBuilder.py

```diff
@@ -143,7 +144,69 @@ def _build_response_payload(self, response: ResponseInfoWrapper) -> dict:
             },
         }
 
-    # always returns a dict with some of these fields: text, mimeType, params}
+    def _get_query_string(self, request):
+        """Helper function to get the query string for a request, translating fields from
+        either a Werkzeug Request object or a Django WSGIRequest object.
+
+        Args:
+            request (Request): Request object containing the request information, either
+                a `werkzeug.Request` or a `django.core.handlers.wsgi.WSGIRequest`.
+
+        Returns:
+            str: Query string, for example "field1=value1&field2=value2"
+        """
+        if hasattr(request, "query_string"):
+            # works for Werkzeug request objects only
+            result = request.query_string
+        elif "QUERY_STRING" in request.environ:
+            # works for Django, and possibly other request objects too
+            result = request.environ["QUERY_STRING"]
+        else:
+            raise Exception(
+                "Don't know how to retrieve query string from this type of request"
+            )
+
+        if isinstance(result, bytes):
+            result = result.decode("utf-8")
+
+        return result
+
+    def _build_base_url(self, request):
+        """Helper function to get the base URL for a request (full URL excluding the
+        query string), translating fields from either a Werkzeug Request object or a
+        Django WSGIRequest object.
+
+        Args:
+            request (Request): Request object containing the request information, either
+                a `werkzeug.Request` or a `django.core.handlers.wsgi.WSGIRequest`.
+
+        Returns:
+            str: Query string, for example "https://api.example.local:8080/v1/userinfo"
+        """
+        if hasattr(request, "base_url"):
+            # Werkzeug request objects already have exactly what we need
+            return request.base_url
+
+        scheme, host, path = None, None, None
+
+        if "wsgi.url_scheme" in request.environ:
+            scheme = request.environ["wsgi.url_scheme"]
+
+        if hasattr(request, "_get_raw_host"):
+            # Django request objects already have a properly formatted host field
+            host = request._get_raw_host()
+        elif "HTTP_HOST" in request.environ:
+            host = request.environ["HTTP_HOST"]
+
+        if "PATH_INFO" in request.environ:
+            path = request.environ["PATH_INFO"]
+
+        if scheme and path and host:
+            return f"{scheme}://{host}{path}"
```

Comment:
The preceding `f` here like some sprintf thing?

Reply:
Yup, exactly 😄